### PR TITLE
Improvements to LBFGS Solver.

### PR DESCRIPTION
**Pull request by hoytak**
Fix for https://github.com/apple/turicreate/issues/1356.

This fix addresses two issues:

- Occasionally the value jumped after a step.  While the algorithm
  recovered, this could lead to a suboptimal value being reported as the
  final solution.  This PR adds in a bracketing back-tracking routine
  to ensure that every subsequent value is an improvement on the
  previous.

- more_thuente is called too infrequently, causing the algorithm to be
  stuck at an inoptimal step size for two long.  This expands the
  conditions that trigger a step size search to include differences in
  the line search value or steps that could have been much more
  significant.

- The input function was not scaled, leading to non-trivial numerical 
  issues in some cases.

### CODE REVIEW

**Comment by hoytak** [comment]
> @hoytak What is a `more_thuente` routine? And what issue is this fixing?

This is the routine that sets the step size.  It requires a few passes over the data, so we do it as little as possible, but if we don't do it enough it will converge too slowly.  This fix puts in additional conditions under which to call that routine, which helps the convergence.

**Comment by znation** [comment]
@hoytak Reviewed and builds are passing. Is this ready to go in?

File: src/optimization/lbfgs.cpp [hunk 1]
```diff
@@ -137,20 +135,17 @@ void lbfgs_solver::setup(
...
     auto it = opts.find(n);
     return (it != opts.end()) ? it->second : default_solver_options.at(n);
   };
 
-
   lbfgs_memory_level = get_param("lbfgs_memory_level");
   convergence_threshold = get_param("convergence_threshold");
-  m_status.step_size = 1; // get_param("step_size");
-
+  m_status.step_size = 1.0;  // / std::sqrt(num_variables);
```
Comment (znation) [inline src/optimization/lbfgs.cpp]: What's with the comment? The value doesn't seem to match the comment...

**Comment by srikris** [comment]
LGTM; Have we tested this on many datasets. Just to make sure we haven't regressed.

File: src/optimization/lbfgs.cpp (hunk 1 above)
Comment (hoytak) [reply src/optimization/lbfgs.cpp]: Bad comment.  It's set automatically now.

**Pull request merged**